staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go

```diff
@@ -139,3 +140,84 @@ type CustomResourceDefinitionList struct {
 	// Items individual CustomResourceDefinitions
 	Items []CustomResourceDefinition
 }
+
+// CustomResourceValidation is a list of validation methods for CustomResources.
+type CustomResourceValidation struct {
+	// OpenAPIV3Schema is the OpenAPI v3 schema to be validated against.
+	OpenAPIV3Schema *JSONSchemaProps
```

Comment:
How big are these normally?  What max size limit is imposed (our 1m value limit)?

Reply:
We cannot rely on an external library to basically do our validation and to follow our api guidelines. So yes, it's about compatibility. We were discussing this in the proposal up and down and voted for this solution for essentially the reason of having control. Still I don't get the argument. The complexity is not big, look at the validations. It's completely straight forward. What is left is a proto marshaller.

Reply:
What is discussed before was possibility of including the external type in our API (that I agree not a good idea) but the suggestion here is to use rawExtension. What we unmarshall it to is totally an implementation detail not a design one. So two things here:

- What exposed in our API! Can be RawExtension with a comment that this is standard Vxxx. Or can be our defined type (that may be slightly different than standard)
- How we implement it internally. With RawExtension we can unmarshall it to an internal type or go-openapi type. It is not user faced. With our defined type exposed, no options here.

The reason I think rawExtension is a good idea is that we don't/can't apply our API standards to an already defined standard (without changing the standard, e.g. StringOrStringArray). Having control over spec means we can diverge from the standard and I don't think having that power is a good thing.

Having RawExtension can also let us support different version of Spec without adding new fields. The standard has the version in it, so we can just look at it first before unmarshall it to the right type.

Reply:
We discussed RawExtension as well. It does not make a big difference how the unmarshaling works, whether via RawExtenions or our own types. In the end we define an external API surface we have to support and RawExtensions are an implementation detail to delegate the unmarshaling. It does not solve any of the API validation and stability problems. Moreover, we know that go-openapi's validation of JSON Schema is weak at best. IMO we cannot trust it for this purpose and have no control over it. We also discussed to use the `$scheme` field for the version before and decided against it (that's why we have `OpenAPIv3` now btw).

In general, we have been through all of these considerations in the proposal and yet again for some here in the implementation PR. I would very much appreciate participation in the proposal process for these kinds of discussions. The proposal was posted to the relevant lists and people were CC'ed on the proposal PR and this implementation PR very early on. It's pretty tiresome to reopen these discussions again and again so late in the process when the PR is ready.

Reply:
You are completely right. I am sorry that we didn't catch this in the design phase. But I am just worried that we add something here that we would stuck supporting while there is, to my opinion, a better option. Using RawExtension is not an implementation detail here and should not impose a lot of changes (we can still keep our type just make it internal, if we want to use go-openapi or not is at another level that can be changed or kept later without changing the surface of our API). I would leave this to @lavalamp to give his final opinion.

Reply:
> If we don't actually use the schema and pass it through to other libraries, then do we really want to bless it by writing it up in our API language? The only reason I see to do that is to ensure backwards compatibility, and I'm not sure if that's persuasive.

The compatibility and control over the serialization both matter since a change in a backing API can easily produce breaking API changes in your exposed (and already serialized!) objects.  In OpenShift, we found this out the hard way by "hiding" a docker library behind a raw serialization.  We could still deserialize the code, but it not longer worked.  If we want it in our API, with our stability guarantees, saying "the backing library changed so that json you sent us broke, it wasn't *really* part of our API" isn't a stance that is consistent with the API guarantees that we have.

We must control the serialization.